packages/angular_devkit/schematics_cli/bin/schematics.ts

```diff
@@ -68,192 +45,248 @@ function usage(exitCode = 0): never {
  * @param str The argument to parse.
  * @return {{collection: string, schematic: (string)}}
  */
-function parseSchematicName(str: string | null): { collection: string, schematic: string } {
+function parseSchematicName(str: string | null): { collection: string, schematic: string | null } {
   let collection = '@schematics/schematics';
 
-  if (!str || str === null) {
-    usage(1);
-  }
-
-  let schematic: string = str as string;
-  if (schematic.indexOf(':') != -1) {
+  let schematic = str;
+  if (schematic && schematic.indexOf(':') != -1) {
     [collection, schematic] = schematic.split(':', 2);
-
-    if (!schematic) {
-      usage(2);
-    }
   }
 
   return { collection, schematic };
 }
 
 
-/** Parse the command line. */
-const booleanArgs = [
-  'allowPrivate',
-  'debug',
-  'dry-run',
-  'force',
-  'help',
-  'list-schematics',
-  'verbose',
-];
-const argv = minimist(process.argv.slice(2), {
-  boolean: booleanArgs,
-  default: {
-    'debug': null,
-    'dry-run': null,
-  },
-  '--': true,
-});
-
-/** Create the DevKit Logger used through the CLI. */
-const logger = createConsoleLogger(argv['verbose']);
-
-if (argv.help) {
-  usage();
+export interface MainOptions {
+  args: string[];
+  stdout?: ProcessOutput;
+  stderr?: ProcessOutput;
 }
 
-/** Get the collection an schematic name from the first argument. */
-const {
-  collection: collectionName,
-  schematic: schematicName,
-} = parseSchematicName(argv._.shift() || null);
-const isLocalCollection = collectionName.startsWith('.') || collectionName.startsWith('/');
+// tslint:disable-next-line:no-big-function
+export async function main({
+  args,
+  stdout = process.stdout,
+  stderr = process.stderr,
+}: MainOptions): Promise<0 | 1> {
+
+  /** Parse the command line. */
+  const booleanArgs = [
+    'allowPrivate',
+    'debug',
+    'dry-run',
+    'force',
+    'help',
+    'list-schematics',
+    'verbose',
+  ];
+  const argv = minimist(args, {
+    boolean: booleanArgs,
+    default: {
+      'debug': null,
+      'dry-run': null,
+    },
+    '--': true,
+  });
+
+  /** Create the DevKit Logger used through the CLI. */
+  const logger = createConsoleLogger(argv['verbose'], stdout, stderr);
+
+  /**
+ * Show usage of the CLI tool, and exit the process.
+ */
+  function usage(exitCode: 1 | 0 = 0): 1 | 0 {
```

Comment:
Usage should probably be outside of main, which will bring this function under the size limit for the no-big-function lint rule.

Reply:
@hansl Updated as requested